### DISCUSSION TITLE
Does eap-dynamic -> (eap-radius or eap-tls) make sense for a break-glass scenario?

### DISCUSSION THREAD

**Discussion by lowjoel** [strongswan/strongswan#1574] [Q&A] [answered]
Hello, I've been trying to think about how to handle authentication of my IPsec tunnel if my normal RADIUS server (via eap-radius) happens to be offline (e.g I am not physically at my network, and the RADIUS server needs restarting). I was thinking of something along the lines of using the eap-dynamic plugin to first try eap-radius as a first round, and if that fails and I present a certificate that matches a trusted CA on the router, I still get to establish the same tunnel.

I know the best way to solve this is to have multiple RADIUS servers and have them fail over to each of them, but I'm trying to think of some sort of "break glass"/worst case scenario here, hence thinking of a much more minimal solution. eap-dynamic does not recognise eap-radius as a possible mechanism (I'm guessing because eap-radius itself has more than one way of handling auth?)

Are there other ways of accomplishing what I want here? Should I try to hack around eap-dynamic to allow something like:

```
eap-dynamic {

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    load = yes

    # Prefer peer's proposed EAP methods.
    prefer_user = no

    # The preferred EAP method(s) to be used.
    preferred = eap-radius,eap-tls

}
```

I think by modifying the code that enumerates all registered eap auth plugins to also yield eap-radius should do the trick. Does it make sense or is this entirely absurd?

**Comment by tobiasbrunner** [comment, marked as answer]
> if that fails and I present a certificate that matches a trusted CA on the router, I still get to establish the same tunnel.

How? That's a completely different authentication mechanism (certificate vs. EAP authentication). Unless, I misunderstood what you mean.

Also, the _eap-dynamic_ plugin can not recover from actually failing EAP methods. If the RADIUS server is not reachable, the _eap-radius_ plugin will report an EAP failure, which the _eap-dynamic_ plugin would just forward to the EAP framework. The purpose of the [_eap-dynamic_ plugin](https://docs.strongswan.org/docs/5.9/plugins/eap-dynamic.html) is to allow clients to select the EAP method that's used via EAP-Nak. If the selected method fails, there will not be a try with another method.

> I know the best way to solve this is to have multiple RADIUS servers and have them fail over to each of them

I think that's the only reasonable way to handle this. Why use a RADIUS server in the first place if you can just terminate the authentication on the VPN gateway anyway (depending on the EAP method you'd also have to duplicate or have access to the client credentials etc.)?

**Reply by lowjoel** [reply]
Cool, thanks for clearing that up. I'm still trying to figure my way around this and learn 😄 

> Why use a RADIUS server in the first place if you can just terminate the authentication on the VPN gateway anyway?

I was thinking of the RADIUS server being the business-as-usual case where everyone has the ability to connect and use. However, in the event of things going wrong - electrical failure, fire, network disconnection of the RADIUS infrastructure, hence "break glass" - a smaller subset of people (e.g the infra engineers like me) can still connect and restore service for the rest of the org by bringing the RADIUS infra back up. This can save making a trip down to the affected location physically.

I guess this could also be accomplished by setting up a separate IKE config which then establishes a similar phase2 tunnel? 🤔

**Reply by tobiasbrunner** [reply]
> I guess this could also be accomplished by setting up a separate IKE config which then establishes a similar phase2 tunnel? thinking

Yes, you could do that. However, that second config can only use EAP authentication if you can distinguish them via IKE identities or IP addresses (strongSwan currently doesn't support connection switching based on EAP-Identities). Otherwise, I'd recommend you just use IKEv2 certificate authentication for that second config.

**Reply by lowjoel** [reply]
Right. Thanks for setting me straight conceptually! 😁 I'll give it a shot.
